src/parser/types.rs

```diff
@@ -200,7 +195,7 @@ pub enum TyKind {
 impl TyKind {
     pub fn match_expected(&self, expected: &TyKind) -> bool {
         match (self, expected) {
-            (TyKind::BigInt, TyKind::Field) => true,
+            (TyKind::Field { .. }, TyKind::Field { .. }) => true,
```

Comment:
If `match_expected` is a relaxed version of `same_as`, and the lhs and rhs doesn't matter, then it seems the following aligns with the intention:
```rust
(TyKind::Field { .. }, TyKind::Field { .. }) => true,
```

In practice, this relaxed check for field type seems necessary. Here are the cases where the `match_expected` and `same_as` are applied to check for the field type.

Assignment case:
```rust
fn main(pub xx: Field) {
  let mut yy = 2; // constant
  yy = xx; // should allow assigning a var value
}
```

Argument case:
```rust
allow either constant or non constant argument
fn add(xx: Field, yy: Field) -> Field {
  return xx + yy; // should allow binary ops over the mix of constants or non constants
}

// yy only allows constant value. 
fn add_const(xx: Field, const yy: Field) -> Field {
 return xx + yy; 
}

fn main(pub xx: Field) {
  // first arg is var, while 2 is a constant
  add(xx, 2);
}
```

Custom type declaration case:
```rust
Thing {
  xx: Field // allow either constant or not
}
fn main(pub xx: Field) {
  // based on var
  let thing1 = Thing {xx: xx};
  // based on constant
  let thing2 = Thing {xx: 2};
}
```

It is indeed getting a bit weird to see the `same_as` function also allows relaxed checking for the Field type though.

Maybe we should rename the `match_expected` to `match_generic`.

Or maybe we should add a bunch of functions to specify individually for each of these cases, instead of using an unified rule for all the cases listed above. For example:
- check_assignment(lhs_typ, rhs_typ)
- check_argument(observed_typ, expected)
- check_return(observed, expected)

This may make it clearer the context in which what exceptions can be made, such as for the Field case.